Mount the external journal device and write accessible extent metadata

jason50123 opened the pull request:
This pull request currently adds the parse_option in fill super to mount the external journal device to simplefs. It also writes the currently accessible extent metadata to the journal during write operations. The image below demonstrates that it is already possible to mount the external journal device in the v6.8 environment.

- Implement external journal device initialization
- Add journal to write function

![螢幕快照 2024-06-20 10-17-03](https://github.com/sysprog21/simplefs/assets/57582802/2fbd3355-9a13-486b-a6d7-eed80fe1e8b4)

jason50123 committed:
Implement external journal device initialization

During fill super, call simplefs_parse_options to find the external
device path and set it as the journal device. Additionally, add the
`make journal` parameter in the Makefile to create an external journal
device.

jason50123 committed:
Add journal to write function

In the current design of simplefs, we can only access the extent
metadata and perform the 'journal' action on it. To write inode and
other metadata, we might need to emulate `ext4_dirty_inode`.
This method primarily uses the content of struct ext4_iloc to find the
correct location of the inode on disk and records it in the journal.
However, in our current situation, simply using `mark_inode_dirty` to
handle dirty inodes does not allow us to obtain the inode's location,
thus preventing us from recording it in the journal.

But currently, the idea is that we might create a new buffer head, use
`jbd2_journal_get_create_access` to put the dirty inode in it, and then
use `jbd2_journal_dirty_metadata` to write it to the journal.

jserv commented:
Avoid uploading screenshots that contain only text. Instead, use Markdown formatting to present command lists, source code, and any text-based content. This approach is more accessible for individuals with visual impairments.

jserv commented on file.c:
```diff
@@ -146,6 +169,16 @@ static int simplefs_write_begin(struct file *file,
     if (nr_allocs > sbi->nr_free_blocks)
         return -ENOSPC;
 
+    // handle journal start here
```
Always use C-style comments.

jserv commented on super.c:
```diff
@@ -196,6 +201,251 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
 }
 
+/* journal relation code */
+
```
Remove one blank line.

jserv commented on super.c:
```diff
@@ -196,6 +201,251 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+    *j_start = sb_block;
+
+    /*
+     *Fix me: This number is calculated by dividing the external device capacity
```
Specify "FIXME" instead of "Fix me."

jserv commented on super.c:
```diff
@@ -196,6 +201,251 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+
+    /*
+     *Fix me: This number is calculated by dividing the external device capacity
+     *(8MB) by the external device block size (4096). Currently, I don't know
```
Improve your English writing.

jserv commented on super.c:
```diff
@@ -196,6 +201,251 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+
+    pr_info("simplefs_parse_options: parsing options '%s'\n", options);
+
+    while ((p = strsep(&options, ",")) != NULL) {
```
Simply rewrite as `while ((p = strsep(&options, ","))) {`

jserv reviewed (changes requested):
Append the test suite to reflect journaling support.

jserv commented:
Drop unnecessary `pr_info` statements.

jserv commented:
Fix compatibility with Linux v6.5:
```
simplefs/super.c: In function ‘simplefs_get_journal_blkdev’:
simplefs/super.c:221:19: error: implicit declaration of function ‘bdev_open_by_dev’; did you mean ‘blkdev_get_by_dev’? [-Werror=implicit-function-declaration]
  221 |     bdev_handle = bdev_open_by_dev(
      |                   ^~~~~~~~~~~~~~~~
      |                   blkdev_get_by_dev
simplefs/super.c:222:48: error: ‘BLK_OPEN_RESTRICT_WRITES’ undeclared (first use in this function)
  222 |         jdev, BLK_OPEN_READ | BLK_OPEN_WRITE | BLK_OPEN_RESTRICT_WRITES, sb,
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~
```

jason50123 committed:
Merge branch 'sysprog21:master' into external-journal

jason50123 committed:
Add support for loading external journal in v6.5

Based on feedback from previous pull requests, addressed the mentioned
issues and ensured that the related external device journal loading
function operates correctly in kernel version 6.5.

jason50123 committed:
Merge branch 'external-journal' of github.com:jason50123/simplefs into external-journal

jason50123 commented:
The latest commit has addressed the mentioned issues. 
Evidence of the journal functioning correctly will be provided later.

jserv reviewed (changes requested):
Use `git rebase -i` to rework the commits.

jserv commented on super.c:
```diff
@@ -332,8 +401,7 @@ static int simplefs_load_journal(struct super_block *sb,
     int err = 0;
-    pr_info("simplefs_load_journal: loading journal from device number %lu\n",
-            journal_devnum);
+
```
Remove this blank line.

jserv reviewed (changes requested):
Ensure that journal device can be mounted for Linux kernels prior to v6.5. At least, tweak for version 6.1, 5.15, and 5.10.
See https://kernel.org/ for LTS.

jason50123 commented:
@jserv Could you please clarify if you would like me to rebase all the commits into a single commit, or if there is a specific way you prefer me to rework them? Thank you!

jserv commented:
> @jserv Could you please clarify if you would like me to rebase all the commits into a single commit, or if there is a specific way you prefer me to rework them? Thank you!

There should be only one commit rebasing the latest `master` branch for reviewing. See https://git-scm.com/docs/git-rebase
